File: _designs_001/Implement Trie (Prefix Tree)/Trie.java
public class Trie {
    private TrieNode root;

    public Trie() {
        root = new TrieNode();
    }

    public void insert(String s) {
        TrieNode p = root;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (p.children[c - 'a'] == null)
                p.children[c - 'a'] = new TrieNode();
            p = p.children[c - 'a'];
        }
        p.end = true;
    }

    public boolean search(String key) {
        TrieNode p = traverse(key);
        return p == null ? false : p.end;
    }
    
    public boolean startsWith(String prefix) {
        return traverse(prefix) != null;
    }
    
    private TrieNode traverse(String s) {
        TrieNode p = root;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (p.children[c - 'a'] == null)
                return null;
            p = p.children[c - 'a'];
        }
        return p;
    }
}

File: _designs_001/Implement Trie (Prefix Tree)/Trie.rb
class Trie
  def initialize
    @root = TrieNode.new
  end

  def insert(s)
    p = @root
    (0...s.length).each do |i|
      p.children[s[i]] = TrieNode.new if !p.children.key?(s[i])
      p = p.children[s[i]]
    end
    p.last = true
  end

  def search(key)
    p = traverse(key)
    p.nil? ? false : p.last
  end

  def startsWith(prefix)
    !traverse(prefix).nil?
  end

  def traverse(s)
    p = @root
    (0...s.length).each do |i|
      return nil if !p.children.key?(s[i])
      p = p.children[s[i]]
    end
    p
  end
end